lib/telemetry_registry.ex

```diff
@@ -0,0 +1,274 @@
+  not limited to the typespec format, i.e. you can include comments or example values. The objective
+  is to inform users what is available.
+  """
+  @type event_metadata() :: :telemetry_registry.event_metadata()
+
+  @typedoc """
+  A map of event definition meta for an event containing the event, measurements, and metadata descriptions
+  if the event was declared with an event definition. Otherwise, this value will be an empty map.
+  """
+  @type event_meta() :: :telemetry_registry.event_meta()
+
+  @typedoc """
+  A list of spannable events known to the registry in the format of `{event_prefix, event_suffixes}`. For
+  example, given events `[:my_app, :request, :start], [:my_app, :request, :stop], [:my_app, :request, :exception]`
+  a span can be created from the `:start` -> `:stop` or the `:start` -> `:exception` events. These are aggregated
+  as a spannable event `{[:my_app, :request], [:start, :stop, :exception]}`.
+  """
+  @type spannable_event() :: :telemetry_registry.spannable_event()
+
+  defmacro __using__(_opts) do
+    quote do
+      import unquote(__MODULE__), only: [telemetry_docs: 0, telemetry_event: 1]
+
+      Module.register_attribute(__MODULE__, :telemetry_event,
+        accumulate: true,
+        persist: true
+      )
+    end
+  end
+
+  @doc """
+  Declares a telemetry event. Accepts a telemetry event definition `t:event_definition/0` or telemetry
+  event name `t:event/0`. A full definition is required for complete documentation and is highly encouraged.
+  """
+  defmacro telemetry_event(event) do
+    quote do
+      @telemetry_event unquote(event)
+    end
+  end
+
+  @doc """
+  Generates telemetry event documentation formatted in Markdown for use in your documentation.
+
+  ```
+  ## Telemetry
+
+  \#{telemetry_docs()}
+  ```
+  """
+  defmacro telemetry_docs do
+    quote do
+      TelemetryRegistry.__docs__(__MODULE__)
+    end
+  end
+
+  @doc false
+  def __docs__(module) do
+    docs_for(module)
+  end
+
+  @doc """
+  Generate telemetry event documentation formatted in Markdown for a given module.
+  """
+  @spec docs_for(module()) :: String.t()
+  def docs_for(module) do
+    get_events(module)
+    |> Enum.map(&format_event/1)
+    |> IO.iodata_to_binary()
+  end
+
+  defp format_event(event) when is_map(event) do
+    """
+    * `#{inspect(event[:event])}`
+      * Description: #{event[:description]}
+      * Measurements: `#{event[:measurements]}`
+      * Metadata: `#{event[:metadata]}`
+
+    """
+  end
+
+  defp format_event(event) when is_list(event) do
+    """
+    * `#{inspect(event)}`
+
+    """
+  end
+
+  defp get_events(module) do
+    try do
+      Module.get_attribute(module, :telemetry_event, [])
+    rescue
+      _ ->
+        module.__info__(:attributes)
+        |> Keyword.get_values(:telemetry_event)
+        |> Enum.map(fn [event] when is_map(event) -> event
+        event -> event
+      end)
```

Comment:
```suggestion
        |> Enum.map(fn
          [event] when is_map(event) -> event
          event -> event
        end)
```

Reply:
Good call. It can also be a `List.flatten` call.

Reply:
I think flatten won't work because it will flatten multiple plain events to one list of atoms, e.g. `List.flatten([[:a, :b, :c], [:d, :e, :f], [%{}])` results in `[:a, :b, :c, :d, :e, :f, %{}]`